### ISSUE TITLE
Meros doesn't support chain reorganizations.

### ISSUE THREAD

**Issue by kayabaNerve** [MerosCrypto/Meros#144]
This is a protocol violation, which traditionally do not get issues unless an existing feature violates the existing protocol. The reason chain reorganizations gets an issue is because of its scope and because it's part of the DT4 milestone.

Chain reorganizations can be defined as being composed of six parts:

- Blockchain reverts.
- State reverts.
- Epochs reverts.
- Consensus reverts.
- Transactions reverts.
- Detecting alternative chains with more work and deciding to sync them instead.

While the first three can be grouped under the umbrella of "Merit reverts", I feel it's optimal to separate them due to the fact they are separate systems, just systems grouped with a master object and common database.

State reverts are notable because they're already implemented. That said, they don't affect the database at all. They're meant to acquire historical data, not to support chain reorganizations.

Blockchain reverts are simply defined as deleting Blocks, including their difficulties, new holders, Merit changes, and reference data. Of course, the actual implementation isn't that simple.

Epochs reverts are special because Epochs don't have any interaction with the Database. While inefficient, as long as Epochs reloading works, we can simply reload the Epochs once the Merit DB is reverted to acquire a valid copy.

Consensus reverts are more complicated. The Consensus DAG notably contains Transaction Statuses, but effectively contains Send Difficulties, Data Difficulties, Gas Prices, and Merit Removals. Despite being stored on the Blockchain, we still need to remove these from the Consensus DAG. This means we need to revert the Consensus DAG before reverting the Blockchain.

Transactions DAG reverts are the final piece of the puzzle and may be the easiest. The Transactions DAG doesn't actually keep track of any Merit or any Verifications. Reverting it is defined as pruning Mint branches, along with updating "spendable". While we don't track if a Transactions is verified, when one becomes verified, we get notified as to mark its outputs as usable in a Send Transaction.

We implemented pruning in https://github.com/MerosCrypto/Meros/commit/cf3a9f19197db558123e7492510d93ee6ac06841, as well as exemplifying how to use it. Currently, the Transactions DAG only supports pruning a single Transaction, while the Consensus DAG implements an algorithm to prune a branch. As we'll now call it from multiple places, it may be worth it to move the branch prune algorithm from Consensus to Transactions.

Detecting alternative chains has a few valid options. I'm personally in favor of syncing every BlockHeader on the alt chain, verifying its work, and then if the work is greater, switching to it. If the BlockBodies end up being invalid, we mark the header as junk and look for the longest chain again. It allows us to keep a single instance of the Merit, Consensus, and Transactions variables.

**Comment by kayabaNerve** [comment]
The above algorithm implements Blockchain reversions which do affect the Database. The Blockchain revert algorithm also calls the State revert to update its cached values. Combined with reloading the Epochs, it does successfully work as Merit reversion code except for one factor: MeritRemovals.

```
template BLOCK_REMOVALS(
    blockNum: int
): string =
    blockNum.toBinary(INT_LEN) & "r"

template HOLDER_REMOVALS(
    nick: uint16
): string =
    nick.toBinary(NICKNAME_LEN) & "r"
```

We index MeritRemovals both by Block and by holder. While deleting the first is one line I should have included in that commit, the second is not. As the BlockchainTest doesn't have any Elements in it, we can't accurately verify it.

**Comment by kayabaNerve** [comment]
The main trick with the Transactions DAG, as specified above, is manipulating spendable.

In my opinion, we should only prune Transactions from the Transactions DAG while reverting. The handling of verified/unverified should happen with the Consensus DAG. In order to prevent loss of Verifications, all Verifications of unfinalized Transactions should have their signatures cached.

This will affect how we currently handle checking if a holder had their Verification archived. We currently check if their signature is still cached to find out.

When the Blockchain is reverted to the point a finalized Transaction becomes unfinalized, and it loses Merit via loss of Verifications, if it's below the node threshold, it should be unverified. This mimics what would happen in real time. Thankfully, this is extremely unlikely thanks to checkpoints.

**Comment by kayabaNerve** [comment]
The Transactions tree prune cannot be combined with the Consensus prune as Consensus needs to know about every pruned Transaction to respond accordingly.

**Comment by kayabaNerve** [comment]
https://github.com/MerosCrypto/Meros/pull/149 finished Transactions reversion. What's left is:

- Consensus reverts.
- Detecting alternative chains with more work and deciding to sync them instead.

**Comment by kayabaNerve** [comment]
Apparently Transactions reversions aren't perfect.

https://github.com/MerosCrypto/Meros/issues/150

**Comment by kayabaNerve** [comment]
In order to not reimplement all the Transactions reversion code, as the Consensus 'DAG' lives on top of the Transactions DAG, the optimal flow is:

```
consensus.revert(merit.blockchain, transactions, height)
transactions.revert(merit.blockchain, height(
merit.blockchain.revert(merit.state, height)
merit.epochs = newEpochs()
consensus.postRevert(merit.blockchain, transactions)
```

There's definitely optimizations to make with the Epochs, Transactions, and Conensus DAGs however. All three don't bother reverting in some way, instead just rebuilding their contents from the database in an accurate manner.

The Epochs rebuild all of themselves. The Transactions DAG rebuilds its cache. The Consensus DAG rebuilds the SpamFilters, and close/unmentioned tables.

**Comment by kayabaNerve** [comment]
Consensus Reversion has its own branch due to the scope of its implementation, testing, and eventual bug fixing.

**Comment by kayabaNerve** [comment]
The existing ConsensusRevert test covers most cases and is mostly passing. That said, it uses 750 MB (fine, expected as it keeps a copy of everything in RAM, sometimes in duplicate or triplicate) until consensus.revert is called. Reverting multiple blocks at once, not one-by-one which is fine, causes the RAM to spike into the tens of GB.

**Comment by kayabaNerve** [comment]
Re-opening as I clicked the wrong button by accident.

**Comment by kayabaNerve** [comment]
Memory problems have been solved in https://github.com/MerosCrypto/Meros/commit/840c438ae6f9ebb26f5d01f9c121062e1745fdf3.

https://github.com/MerosCrypto/Meros/commit/c397056d27af256008053015dc5f569dbbf903e4 details the work remaining before the `Consensus-Reversions` branch can be merged. One of the checks (whether not a Transaction is verified) is also currently failing, yet will be fixed shortly.

**Comment by kayabaNerve** [comment]
The ConsensusReversions test now passes the 3.9 GB mark, a number I'm not happy with. While it is reverting 105 blocks at once when this happens, we may want to consider batching Consensus reversions in groups of 10 to decrease the required amount of memory.

**Comment by kayabaNerve** [comment]
The test finally passes with its existing checks on a repeatable basis, thankfully. That said, the amount of memory is a requirement due to the amount of Transaction created. Without lowering the thoroughness of the test, we'll have to deal with it.

**Comment by kayabaNerve** [comment]
I ran the Consensus Revert test, as of the current pushed Consensus Revert branch. It ran overnight, for roughly 18 hours total, without failing.

I've been keeping a list of what's left to do before this branch can be merged.

- Extend the existing test by loading the Send/Data diffs for every holder and verifying their accuracy, instead of just reloading the SpamFilters and comparing them.

- Add a new test to verify `competing`. `competing` should always be untouched, even if the competitor was pruned. The fact it ever had a competitor means we should only trust if its verified when finalization occurs.

- `beaten` should be if its parents were beaten, not if it was beaten. We can write up a few specific cases to verify this behavior.

- Verify deleteMeritRemoval reopens the nonce for usage for a holder's chosen difficulty.

- Verify deleteMeritRemoval only deletes one nonce. If other MeritRemovals also involve the nonce, the nonce should remain blacklisted.

- Verify we preserve a copy of TXs about to be pruned which are proof of a pending VC MR. This should be a Python test.

I don't want to extend the existing test with the `competing` or `beaten` checks because it requires creating competing Transactions, which are a royal pain to maintain for a variety of reasons. Creating smaller, dedicated tests, will be optimal.

**Comment by kayabaNerve** [comment]
- I just pushed a commit extending the test to verify the difficulties for every holder.

- Due to the need for a larger set of tests around `competing`, and the fact the behavior we're testing is the lack of behavior, I'm fine with not including the relevant tests in this PR, or requiring the tests to be submitted for this issue to be closed.

- `beaten` also needs a larger set of tests, yet we also need to test proper implementation of reversions. Due to the existing large scope and size of the branch, I'd like to merge before those tests are completed yet they must be done before this issue is closed.

- I'm working now on the MR tests. We should be able to test them all in Python rather easily.

**Comment by kayabaNerve** [comment]
To update this, as I haven't in a while, reversions were completed with the merging of the above branch.

The next step was detecting if a chain with more work existed. This is done by summing the difficulties. As we used a 256-bit difficulty, this was an expensive operation. In order to alleviate this expense, difficulty was redone to be 64-bit for the Blockchain and 32-bit for the Transaction DAG (as the difficulty was updated and standardized across the codebase).

I then added basic reorg code to detect longer chain and revert to the last common Block. I didn't write code to add all the Blocks between the last common Block and the new tip That said, if the reorg depth was 1, the code worked IF the Merit Holders remained constant (verified via https://github.com/MerosCrypto/Meros/tree/master/PythonTests/Tests/Merit/ReorgDepthOneTest.py).

I just removed the Merit Holders restriction in https://github.com/MerosCrypto/Meros/commit/a46afc06a57dbd1d663172ece651a7a73f5eaaa1 and am currently working on the Block addition loop, along with extended testing.
